### PR TITLE
Adding Initial Test + Added Document for testing.

### LINKED ISSUE #60: Setting up Initial test + necessary documentation
This issue is to set up an initial test within Jest and beginning adding documentation for testing.

### PR DESCRIPTION
**Pull request by DarrellRichards**
# Pull Request

Added the Initial Test and Document to breakdown some of the testing requirements

Fixes #60 

## Type of change

- [ ] A new feature
- [ ] A bug fix
- [ ] Documentation only changes
- [ ] Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [x] A code change that neither fixes a bug nor adds a feature
- [ ] A code change that improves performance
- [x] Adding missing tests or correcting existing tests
- [ ] Changes that affect the build system or external dependencies
- [ ] Changes to our CI configuration files and scripts
- [ ] Other changes that don't modify src or test files

```release-note
   - Added jwt.middleware.ts test coverage
   - Added TESTING.md document for testing information
```

---

## Checklist

- [x] My code follows the [contributing](../CONTRIBUTING.md) guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Whamo12** [comment]
When I run `npm run test:node` I'm getting
```
 FAIL  src/middleware/jwt.spec.ts (6.924 s)
  jwt.middleware.ts
    ✕ running checkToken should return a 401 with Authorization (4 ms)
    ✕ running checkToken should return a req.user (2 ms)
    ✕ running checkToken should return a 401 without Authorization
    ✕ running checkRefreshToken should return a 401 with refreshToken (1 ms)
    ✕ running checkRefreshToken should return a 401 without refreshToken
    ✕ running checkRefreshToken should return a req.user (1 ms)```

Is this expected?

**Comment by DarrellRichards** [comment]
@Whamo12 

This is not expected, and it gives me an idea of what needs to be added to the testing document.

The reason all the tests are failing is your missing a `.env` with at least the following information in it.

`.env`
```
JWT_KEY="changeMe"
JWT_REFRESH_KEY="changeMe"
```

Couple of ways this can be documented is telling users what needs to be in the .env for testing or providing a `.env.test` that will be used for testing. 


Let me know your thoughts.

**Comment by Whamo12** [comment]
@DarrellRichards Using the env vars for `JWT_KEY` and `JWT_REFRESH_KEY` is unnecessary with tests.  You could easily just replace it with any string such as:

```
  beforeAll(() => {
    token = jwt.sign({ userId: '2222' }, 'someValue', { expiresIn: '15m' });
    refreshToken = jwt.sign({ userId: '2222' }, 'anotherValue', { expiresIn: '8h' });
  });```

File: src/middleware/jwt.spec.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+require('dotenv').config();
```
Comment (Whamo12) [inline src/middleware/jwt.spec.ts:1]: Remove import and replace process.env variables with random string values.

**Comment by DarrellRichards** [comment]
> @DarrellRichards Using the env vars for `JWT_KEY` and `JWT_REFRESH_KEY` is unnecessary with tests. You could easily just replace it with any string such as:
> 
> ```
>   beforeAll(() => {
>     token = jwt.sign({ userId: '2222' }, 'someValue', { expiresIn: '15m' });
>     refreshToken = jwt.sign({ userId: '2222' }, 'anotherValue', { expiresIn: '8h' });
>   });```
> ```

This is interesting, I was always told otherwise. 

I'll make the above changes and get this up to date.

**Comment by Whamo12** [comment]
I understand.  In then end, the middleware is [verifying](https://www.npmjs.com/package/jsonwebtoken#jwtverifytoken-secretorpublickey-options-callback) if the token is valid with a secret or public key.  In production, the key will always be secret.  If you could provide me with additional information as to why we should use the same key from the process env variable in tests that would be appreciated.  Otherwise, I do not see a harm using a random value for testing.

**Comment by DarrellRichards** [comment]
I completely understand where your coming from, I think it more habit than anything else past company I worked for wanted to everything including the `.env`

I do need to see why my tslint is not grabbing those issues you pointed out about the TSLint.

**Comment by Whamo12** [comment]
Running `npm run lint` should catch the issues. Just in case your local longer is not working.

**Comment by DarrellRichards** [comment]
I have to set the `process.env.JWT_KEY and Refresh` in the beforeAll, so that the JWT.verify verify correctly, otherwise tests 2 & 6 will fail for never returning the `req.user`

**Comment by Whamo12** [comment]
Alright, after doing research, I managed 6 passing tests.  I added the path property to the dotenv config.  Update the `require` statement to:
```
require('dotenv').config({ path: '.env' });
```

**Comment by DarrellRichards** [comment]
> Alright, after doing research, I managed 6 passing tests. I added the path property to the dotenv config. Update the `require` statement to:
> 
> ```
> require('dotenv').config({ path: '.env' });
> ```
> 
> Here's the [documentation](https://www.npmjs.com/package/dotenv#path)

So the `beforeAll function` didn't seem to fix the issue? 

The reason I ask because it should work and not require us to add the `dotenv` package. 

Research Article: https://medium.com/@maxcbc/mocking-environment-variables-in-node-js-a17a416e127c



Just making sure I'm on the same page, we're going back to adding the `require('dotenv').config({ path: '.env' })` to the `jwt.spec.ts`?

**Comment by Whamo12** [comment]
The issue is we would have to modify the `checkToken` implementation.  For now, let's use the path property with the dotenv path property.  We can revisit with an enhancement later.

**Comment by Whamo12** [comment]
@DarrellRichards Great job, once again.  I'm going to create a refactor ticket to look into the issue with the JWT token.

**Pull request merged**